Panic when divide by zero

i have an tick script:
```
var db = 'telegraf'
var rp = 'autogen'

var measurement = 'mongo_cluster'

var groupBy = ['customer', 'environment', 'project']

var whereFilter = lambda: TRUE

var name = 'backend_mongodb_cluster_health'

var idVar = name + '-{{.Group}}'

var message = 'MongoDB Cluster on project {{ index .Tags "project" }} is DOWN'

var idTag = 'alertID'

var levelTag = 'level'

var messageField = 'message'

var durationField = 'duration'

var triggerType = 'threshold'

var crit = 50

var data = stream
    |from()
        .database(db)
        .retentionPolicy(rp)
        .measurement(measurement)
        .groupBy(groupBy)
        .where(whereFilter)
    |eval(lambda: 100 * ("total_members" - "active_members") / "total_members")
        .as('value')
        .quiet()

var trigger = data
    |alert()
        .crit(lambda: "value" < crit)
        .message(message)
        .id(idVar)
        .idTag(idTag)
        .levelTag(levelTag)
        .messageField(messageField)
        .durationField(durationField)
        .stateChangesOnly(30s)
        .log('/tmp/mongodb.log')
```
when data go to steam with "total_members" = 0 . it go to panic:
```
panic: runtime error: integer divide by zero
goroutine 290 [running]:
github.com/influxdata/kapacitor/tick/stateful.glob..func51(0xc000fce900, 0xc000de1e90, 0x25317a0, 0xc002935e80, 0x2531b20, 0xc00297aeb0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/evaluation_funcs.go:1037 +0x31c
github.com/influxdata/kapacitor/tick/stateful.(*EvalBinaryNode).eval(0xc002935e40, 0xc000fce900, 0xc000de1e90, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/eval_binary_node.go:254 +0xba
github.com/influxdata/kapacitor/tick/stateful.(*EvalBinaryNode).evaluateDynamicNode(0xc002935e40, 0xc000fce900, 0xc000de1e90, 0x25317a0, 0xc002935e80, 0x2531b20, 0xc00297aeb0, 0x0, 0x0, 0x0, ...)
         /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/eval_binary_node.go:312 +0x330
github.com/influxdata/kapacitor/tick/stateful.(*EvalBinaryNode).eval(0xc002935e40, 0xc000fce900, 0xc000de1e90, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
         /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/eval_binary_node.go:254 +0xba
github.com/influxdata/kapacitor/tick/stateful.(*EvalBinaryNode).EvalInt(0xc002935e40, 0xc000fce900, 0xc000de1e90, 0xc0007df202, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/eval_binary_node.go:231 +0x5a
github.com/influxdata/kapacitor/tick/stateful.(*expression).EvalInt(...)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/expr.go:72
github.com/influxdata/kapacitor/tick/stateful.(*expression).Eval(0xc000fce8a0, 0xc000fce900, 0x2, 0x2, 0x7fa4553f0b30, 0xc0004819e0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/tick/stateful/expr.go:99 +0x22b
github.com/influxdata/kapacitor.(*EvalNode).eval(0xc00291d600, 0xc000fca370, 0x1, 0x1, 0x7fa4553f0c58, 0xc0004819e0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/eval.go:104 +0xe90
github.com/influxdata/kapacitor.(*evalGroup).doEval(0xc000fce8c0, 0x7fa4553f0c58, 0xc0004819e0, 0x7fa4553f0c58)
         /home/ubuntu/go/src/github.com/influxdata/kapacitor/eval.go:210 +0x61
github.com/influxdata/kapacitor.(*evalGroup).Point(0xc000fce8c0, 0x253f040, 0xc0004819e0, 0xb904ec, 0x1dfef20, 0xc000f705d0, 0xc000b2c340)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/eval.go:203 +0x7e
github.com/influxdata/kapacitor/edge.(*timedForwardReceiver).Point(0xc000fce8e0, 0x253f040, 0xc000481950, 0x0, 0x0, 0x2526ee0, 0xc000de1ec0)
         /home/ubuntu/go/src/github.com/influxdata/kapacitor/edge/timed.go:62 +0x62
github.com/influxdata/kapacitor/edge.(*forwardingReceiver).Point(0xc000de1ec0, 0x253f040, 0xc000481950, 0xc000de1e60, 0x0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/edge/forwarding.go:76 +0x4c
github.com/influxdata/kapacitor/edge.(*groupedConsumer).Point(0xc001066240, 0x253f040, 0xc000481950, 0x253f040, 0xc000481950)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/edge/grouped.go:107 +0x151
github.com/influxdata/kapacitor/edge.(*consumer).Consume(0xc000d34440, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/edge/consumer.go:62 +0x38e
github.com/influxdata/kapacitor/edge.(*groupedConsumer).Consume(0xc001066240, 0x2190b14, 0x13)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/edge/grouped.go:51 +0x33
github.com/influxdata/kapacitor.(*EvalNode).runEval(0xc00291d600, 0x0, 0x0, 0x0, 0x2e33, 0x2140540)
       /home/ubuntu/go/src/github.com/influxdata/kapacitor/eval.go:72 +0x1bb
 github.com/influxdata/kapacitor.(*node).start.func1(0xc00291d600, 0x0, 0x0, 0x0)
        /home/ubuntu/go/src/github.com/influxdata/kapacitor/node.go:186 +0x89
created by github.com/influxdata/kapacitor.(*node).start
      /home/ubuntu/go/src/github.com/influxdata/kapacitor/node.go:164 +0x5d
```
Make kapacitord restart over and over again. Even difficult to disable that tickscript.

I know i can workaround by add WHERE condition check if "total_members" !=0 . But it just a trick. All service down just because adding one script seem like unacceptable. 
Can you fix that? make it not panic again?

--- a/eval.go
+++ b/eval.go
@@ -91,8 +91,7 @@ func (n *EvalNode) newGroup() *evalGroup {
 	}
 }
 
-func (n *EvalNode) eval(expressions []stateful.Expression, p edge.FieldsTagsTimeSetter) error {
-
+func (n *EvalNode) eval(expressions []stateful.Expression, p edge.FieldsTagsTimeSetter) (err error) {
 	vars := n.scopePool.Get()
 	defer n.scopePool.Put(vars)
 
@@ -172,7 +171,7 @@ func (n *EvalNode) eval(expressions []stateful.Expression, p edge.FieldsTagsTime
 	}
 	p.SetFields(newFields)
 	p.SetTags(newTags)
-	return nil
+	return
 }
 
 type evalGroup struct {

--- a/integrations/streamer_test.go
+++ b/integrations/streamer_test.go
@@ -6,8 +6,6 @@ import (
 	"encoding/json"
 	"flag"
 	"fmt"
-	"github.com/influxdata/kapacitor/services/discord"
-	"github.com/influxdata/kapacitor/services/discord/discordtest"
 	"html"
 	"io/ioutil"
 	"math/rand"
@@ -22,6 +20,9 @@ import (
 	"testing"
 	"text/template"
 	"time"
+
+	"github.com/influxdata/kapacitor/services/discord"
+	"github.com/influxdata/kapacitor/services/discord/discordtest"
 
 	"github.com/davecgh/go-spew/spew"
 	"github.com/docker/docker/api/types/swarm"
@@ -2689,6 +2690,20 @@ stream
 	}
 
 	testStreamerWithOutput(t, "TestStream_EvalAllTypes", script, 2*time.Second, er, false, nil)
+}
+
+func TestStream_EvalDivisionByZero(t *testing.T) {
+	var script = `
+stream
+	|from()
+		.measurement('data')
+	|eval(lambda: 10/"n")
+		.as('n')
+	|httpOut('TestStream_EvalDivisionByZero')
+`
+
+	testStreamerNoOutput(t, "TestStream_EvalDivisionByZero", script, 2*time.Second, nil)
+
 }
 
 func TestStream_Eval_KeepAll(t *testing.T) {
@@ -12868,6 +12883,7 @@ func testStreamer(
 	<-chan error,
 	*kapacitor.TaskMaster,
 ) {
+	t.Helper()
 	if testing.Verbose() {
 		wlog.SetLevel(wlog.DEBUG)
 	} else {
@@ -13027,6 +13043,7 @@ func testStreamerWithOutput(
 	ignoreOrder bool,
 	tmInit func(tm *kapacitor.TaskMaster),
 ) {
+	t.Helper()
 	clock, et, replayErr, tm := testStreamer(t, name, script, tmInit)
 	defer tm.Close()
 

--- a/integrations/testdata/TestStream_EvalDivisionByZero.srpl
+++ b/integrations/testdata/TestStream_EvalDivisionByZero.srpl
@@ -1,0 +1,3 @@
+dbname
+rpname
+data,t=t1 n=0i 0000003600

--- a/tick/stateful/expr.go
+++ b/tick/stateful/expr.go
@@ -1,6 +1,7 @@
 package stateful
 
 import (
+	"errors"
 	"fmt"
 	"time"
 
@@ -88,50 +89,64 @@ func (se *expression) EvalMissing(scope *Scope) (*ast.Missing, error) {
 	return se.nodeEvaluator.EvalMissing(scope, se.executionState)
 }
 
-func (se *expression) Eval(scope *Scope) (interface{}, error) {
-	typ, err := se.nodeEvaluator.Type(scope)
+func (se *expression) Eval(scope *Scope) (result interface{}, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch r := r.(type) {
+			case string:
+				err = errors.New(r)
+			case error:
+				err = r
+			case fmt.Stringer:
+				err = errors.New(r.String())
+			}
+		}
+	}()
+	var typ ast.ValueType
+	typ, err = se.nodeEvaluator.Type(scope)
 	if err != nil {
 		return nil, err
 	}
 
 	switch typ {
 	case ast.TInt:
-		result, err := se.EvalInt(scope)
+		result, err = se.EvalInt(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	case ast.TFloat:
-		result, err := se.EvalFloat(scope)
+		result, err = se.EvalFloat(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	case ast.TString:
-		result, err := se.EvalString(scope)
+		result, err = se.EvalString(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	case ast.TBool:
-		result, err := se.EvalBool(scope)
+		result, err = se.EvalBool(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	case ast.TDuration:
-		result, err := se.EvalDuration(scope)
+		result, err = se.EvalDuration(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	case ast.TMissing:
-		result, err := se.EvalMissing(scope)
+		result, err = se.EvalMissing(scope)
 		if err != nil {
 			return nil, err
 		}
 		return result, err
 	default:
-		return nil, fmt.Errorf("expression returned unexpected type %s", typ)
+		err = fmt.Errorf("expression returned unexpected type %s", typ)
+		return
 	}
 }

--- a/tick/stateful/expr_dynamic_test.go
+++ b/tick/stateful/expr_dynamic_test.go
@@ -30,6 +30,7 @@ type testCase struct {
 // runDyanmicTestCase is when we want to change the "dyanmism" of
 // a node - type change or value change
 func runDynamicTestCase(t *testing.T, tc testCase) {
+	t.Helper()
 	se := mustCompileExpression(tc.Node)
 
 	for i, expectation := range tc.Expectations {
@@ -50,8 +51,11 @@ func runDynamicTestCase(t *testing.T, tc testCase) {
 			result, err = se.Eval(scope)
 		}
 
-		if err != nil {
-			if expectation.ExpectedError == nil {
+		if err != nil || expectation.ExpectedError != nil {
+			if err == nil {
+				t.Errorf("%s, %s, Iteration: %v: Expected %v error, but got nil", tc.Title, evaluationType, (i + 1), expectation.ExpectedError)
+
+			} else if expectation.ExpectedError == nil {
 				t.Errorf("%s: %s: Iteration %v: Got unexpected error while expecting for result:\n %v\n", tc.Title, evaluationType, (i + 1), err)
 				continue
 			} else if err.Error() != expectation.ExpectedError.Error() {
@@ -67,94 +71,12 @@ func runDynamicTestCase(t *testing.T, tc testCase) {
 }
 
 func TestExpression_BinaryNode_DynamicTestCases(t *testing.T) {
+	t.Run("BinaryNode - EvalBool supports numeric type changes", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
 
-	runDynamicTestCase(t, testCase{
-		Title: "BinaryNode - EvalBool supports numeric type changes",
-
-		Node: &ast.BinaryNode{
-			Operator: ast.TokenGreater,
-			Left: &ast.ReferenceNode{
-				Reference: "value",
-			},
-			Right: &ast.NumberNode{
-				IsFloat: true,
-				Float64: float64(10),
-			},
-		},
-
-		Expectations: []valueExpectation{
-			{IsEvalBool: true, Value: float64(20), ExpectedError: nil, ExpectedResult: true},
-			{IsEvalBool: true, Value: int64(5), ExpectedError: nil, ExpectedResult: false},
-		},
-	})
-
-	runDynamicTestCase(t, testCase{
-		Title: "BinaryNode - EvalNum supports numeric type changes",
-
-		Node: &ast.BinaryNode{
-			Operator: ast.TokenPlus,
-			Left: &ast.ReferenceNode{
-				Reference: "value",
-			},
-			Right: &ast.NumberNode{
-				IsFloat: true,
-				Float64: float64(10),
-			},
-		},
-
-		Expectations: []valueExpectation{
-			{
-				IsEvalNum:      true,
-				Value:          float64(20),
-				ExpectedError:  nil,
-				ExpectedResult: float64(30),
-			},
-			{
-				IsEvalNum:      true,
-				Value:          int64(5),
-				ExpectedError:  errors.New("mismatched type to binary operator. got int + float. see bool(), int(), float(), string(), duration()"),
-				ExpectedResult: nil,
-			},
-		},
-	})
-
-	runDynamicTestCase(t, testCase{
-		Title: "BinaryNode - EvalNum supports numeric value changes",
-
-		Node: &ast.BinaryNode{
-			Operator: ast.TokenGreater,
-			Left: &ast.ReferenceNode{
-				Reference: "value",
-			},
-			Right: &ast.NumberNode{
-				IsFloat: true,
-				Float64: float64(10),
-			},
-		},
-
-		Expectations: []valueExpectation{
-			{
-				IsEvalBool:     true,
-				Value:          float64(20),
-				ExpectedError:  nil,
-				ExpectedResult: true,
-			},
-			{
-				IsEvalBool:     true,
-				Value:          int64(5),
-				ExpectedError:  nil,
-				ExpectedResult: false,
-			},
-		},
-	})
-
-	runDynamicTestCase(t, testCase{
-		Title: "BinaryNode - Nested BinaryNodes can determine correct type",
-
-		Node: &ast.BinaryNode{
-			Operator: ast.TokenAnd,
-			Left: &ast.BinaryNode{
-				Operator: ast.TokenLess,
+			Node: &ast.BinaryNode{
+				Operator: ast.TokenGreater,
 				Left: &ast.ReferenceNode{
 					Reference: "value",
 				},
@@ -163,92 +85,205 @@ func TestExpression_BinaryNode_DynamicTestCases(t *testing.T) {
 					Float64: float64(10),
 				},
 			},
-			Right: &ast.BinaryNode{
+
+			Expectations: []valueExpectation{
+				{IsEvalBool: true, Value: float64(20), ExpectedError: nil, ExpectedResult: true},
+				{IsEvalBool: true, Value: int64(5), ExpectedError: nil, ExpectedResult: false},
+			},
+		})
+
+	})
+	t.Run("BinaryNode - Eval errors on integer division by zero", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
+
+			Node: &ast.BinaryNode{
+				Operator: ast.TokenDiv,
+				Left: &ast.NumberNode{
+					IsInt: true,
+					Int64: 10,
+				},
+				Right: &ast.NumberNode{
+					IsInt: true,
+					Int64: 0,
+				},
+			},
+
+			Expectations: []valueExpectation{
+				{ExpectedError: errors.New("runtime error: integer divide by zero"), IsEvalNum: true},
+			},
+		})
+	})
+
+	t.Run("BinaryNode - EvalNum supports numeric type changes", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
+
+			Node: &ast.BinaryNode{
+				Operator: ast.TokenPlus,
+				Left: &ast.ReferenceNode{
+					Reference: "value",
+				},
+				Right: &ast.NumberNode{
+					IsFloat: true,
+					Float64: float64(10),
+				},
+			},
+
+			Expectations: []valueExpectation{
+				{
+					IsEvalNum:      true,
+					Value:          float64(20),
+					ExpectedError:  nil,
+					ExpectedResult: float64(30),
+				},
+				{
+					IsEvalNum:      true,
+					Value:          int64(5),
+					ExpectedError:  errors.New("mismatched type to binary operator. got int + float. see bool(), int(), float(), string(), duration()"),
+					ExpectedResult: nil,
+				},
+			},
+		})
+	})
+
+	t.Run("BinaryNode - EvalNum supports numeric type changes", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
+
+			Node: &ast.BinaryNode{
 				Operator: ast.TokenGreater,
 				Left: &ast.ReferenceNode{
 					Reference: "value",
 				},
 				Right: &ast.NumberNode{
 					IsFloat: true,
-					Float64: float64(7),
+					Float64: float64(10),
 				},
 			},
-		},
 
-		Expectations: []valueExpectation{
-			{
-				IsEvalBool:     true,
-				Value:          float64(8),
-				ExpectedError:  nil,
-				ExpectedResult: true,
+			Expectations: []valueExpectation{
+				{
+					IsEvalBool:     true,
+					Value:          float64(20),
+					ExpectedError:  nil,
+					ExpectedResult: true,
+				},
+				{
+					IsEvalBool:     true,
+					Value:          int64(5),
+					ExpectedError:  nil,
+					ExpectedResult: false,
+				},
 			},
-			{
-				IsEvalBool:     true,
-				Value:          int64(5),
-				ExpectedError:  nil,
-				ExpectedResult: false,
-			},
-			{
-				IsEvalBool:     true,
-				Value:          int64(11),
-				ExpectedError:  nil,
-				ExpectedResult: false,
-			},
-		},
+		})
 	})
 
+	t.Run("BinaryNode - Nested BinaryNodes can determine correct type", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
+			Node: &ast.BinaryNode{
+				Operator: ast.TokenAnd,
+				Left: &ast.BinaryNode{
+					Operator: ast.TokenLess,
+					Left: &ast.ReferenceNode{
+						Reference: "value",
+					},
+					Right: &ast.NumberNode{
+						IsFloat: true,
+						Float64: float64(10),
+					},
+				},
+				Right: &ast.BinaryNode{
+					Operator: ast.TokenGreater,
+					Left: &ast.ReferenceNode{
+						Reference: "value",
+					},
+					Right: &ast.NumberNode{
+						IsFloat: true,
+						Float64: float64(7),
+					},
+				},
+			},
+
+			Expectations: []valueExpectation{
+				{
+					IsEvalBool:     true,
+					Value:          float64(8),
+					ExpectedError:  nil,
+					ExpectedResult: true,
+				},
+				{
+					IsEvalBool:     true,
+					Value:          int64(5),
+					ExpectedError:  nil,
+					ExpectedResult: false,
+				},
+				{
+					IsEvalBool:     true,
+					Value:          int64(11),
+					ExpectedError:  nil,
+					ExpectedResult: false,
+				},
+			},
+		})
+
+	})
 }
-
 func TestExpression_UnaryNode_DyanmicTestCases(t *testing.T) {
-	runDynamicTestCase(t, testCase{
-		Title: "UnaryNode - EvalNum supports numeric type changes",
+	t.Run("UnaryNode - EvalNum supports numeric type changes", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
 
-		Node: &ast.UnaryNode{
-			Operator: ast.TokenMinus,
-			Node: &ast.ReferenceNode{
-				Reference: "value",
+			Node: &ast.UnaryNode{
+				Operator: ast.TokenMinus,
+				Node: &ast.ReferenceNode{
+					Reference: "value",
+				},
 			},
-		},
 
-		Expectations: []valueExpectation{
-			{
-				IsEvalNum:      true,
-				Value:          float64(20),
-				ExpectedError:  nil,
-				ExpectedResult: float64(-20),
+			Expectations: []valueExpectation{
+				{
+					IsEvalNum:      true,
+					Value:          float64(20),
+					ExpectedError:  nil,
+					ExpectedResult: float64(-20),
+				},
+				{
+					IsEvalNum:      true,
+					Value:          int64(20),
+					ExpectedError:  nil,
+					ExpectedResult: int64(-20),
+				},
 			},
-			{
-				IsEvalNum:      true,
-				Value:          int64(20),
-				ExpectedError:  nil,
-				ExpectedResult: int64(-20),
-			},
-		},
+		})
 	})
 
-	runDynamicTestCase(t, testCase{
-		Title: "UnaryNode - EvalBool supports boolean value changes",
+	t.Run("UnaryNode - EvalBool supports boolean value changes", func(t *testing.T) {
+		runDynamicTestCase(t, testCase{
+			Title: t.Name(),
 
-		Node: &ast.UnaryNode{
-			Operator: ast.TokenNot,
-			Node: &ast.ReferenceNode{
-				Reference: "value",
+			Node: &ast.UnaryNode{
+				Operator: ast.TokenNot,
+				Node: &ast.ReferenceNode{
+					Reference: "value",
+				},
 			},
-		},
 
-		Expectations: []valueExpectation{
-			{
-				IsEvalBool:     true,
-				Value:          bool(true),
-				ExpectedError:  nil,
-				ExpectedResult: bool(false),
+			Expectations: []valueExpectation{
+				{
+					IsEvalBool:     true,
+					Value:          true,
+					ExpectedError:  nil,
+					ExpectedResult: false,
+				},
+				{
+					IsEvalBool:     true,
+					Value:          false,
+					ExpectedError:  nil,
+					ExpectedResult: true,
+				},
 			},
-			{
-				IsEvalBool:     true,
-				Value:          bool(false),
-				ExpectedError:  nil,
-				ExpectedResult: bool(true),
-			},
-		},
+		})
 	})
 }

